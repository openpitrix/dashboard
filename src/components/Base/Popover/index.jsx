import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Tooltip from '../Tooltip';
import styles from './index.scss';

export default class Popover extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    content: PropTypes.oneOfType([PropTypes.node, PropTypes.element]),
    onVisibleChange: PropTypes.func,
    placement: PropTypes.string,
    prefixCls: PropTypes.string,
    showBorder: PropTypes.bool,
    title: PropTypes.string,
    trigger: PropTypes.string,
    visible: PropTypes.bool
  };

  static defaultProps = {
    prefixCls: 'pi-popover',
    trigger: 'click',
    placement: 'bottom',
    showBorder: false
  };

  getOverlay = () => {
    const { prefixCls, title, content } = this.props;

    return (
      <div>
        {title && <div className={`${prefixCls}-title`}>{title}</div>}
        <div className={`${prefixCls}-content`}>{content}</div>
      </div>
    );
  };

  render() {
    const { className, children, ...others } = this.props;

    return (
      <Tooltip
        {...others}
        className={classNames(styles.popover, className)}
        content={this.getOverlay()}
      >
        {children}
      </Tooltip>
    );
  }
}
