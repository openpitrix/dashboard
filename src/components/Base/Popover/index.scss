@import '~scss/vars';

.popover {
  :global {
    .pi-popover-popper {
      min-width: 100px;
      padding: 7px 0;
      font-size: $body-size;
      border-radius: 2px;
      background-color: $N0;
      box-shadow: 0 1px 4px 0 rgba(73, 33, 173, 0.1), 0 4px 8px 0 rgba(35, 35, 36, 0.08);
      border: solid 1px $N30;

      .pi-popover-popper_arrow {
        width: 16px;
        height: 16px;

        &::before, &::after {
          content: "";
          position: absolute;
          width: 0;
          height: 0;
          border-style: solid;
          box-sizing: border-box;
        }
      }

      &[data-placement^="top"] {
        margin-bottom: 4px;

        .pi-popover-popper_arrow {
          bottom: -15px;

          &::before {
            left: 0;
            top: 2px;
            border-width: 8px;
            border-color: $border transparent transparent transparent;
            z-index: 1;
          }
          &::after {
            left: 1px;
            top: 1px;
            border-width: 7px;
            border-color: #fff transparent transparent transparent;
            z-index: 2;
          }
        }
      }
      &[data-placement^="bottom"] {
        margin-top: 4px;
        margin-left: -30px;

        .pi-popover-popper_arrow {
          top: -15px;

          &::before {
            left: 0;
            bottom: 2px;
            border-width: 8px;
            border-color: transparent transparent $border transparent;
            z-index: 1;
          }
          &::after {
            left: 1px;
            bottom: 1px;
            border-width: 7px;
            border-color: transparent transparent #fff transparent;
            z-index: 2;
          }
        }
      }
      &[data-placement^="left"] {
        margin-right: 4px;

        .pi-popover-popper_arrow {
          right: -15px;

          &::before {
            left: 2px;
            border-width: 8px;
            border-color: transparent transparent transparent $border;
            z-index: 1;
          }
          &::after {
            left: 1px;
            top: 1px;
            border-width: 7px;
            border-color: transparent transparent transparent #fff;
            z-index: 2;
          }
        }
      }
      &[data-placement^="right"] {
        margin-left: 4px;

        .pi-popover-popper_arrow {
          left: -15px;

          &::before {
            right: 2px;
            border-width: 8px;
            border-color: transparent $border transparent transparent;
            z-index: 1;
          }
          &::after {
            right: 1px;
            top: 1px;
            border-width: 7px;
            border-color: transparent #fff transparent transparent;
            z-index: 2;
          }
        }
      }
    }

    .operate-menu {
      //margin: 0 -5px;
      a {
        color: $N300;
      }
      li,a,span {
        display: block;
        min-width: 150px;
        padding: 4px 12px;
        height: 32px;
        font-size: 14px;
        line-height: 1.71;
        color: $N300;
        background-color: $N0;
        cursor: pointer;
        transition: all $transition-speed;
        &:hover {
          color: $N500;
          background-color: $N10;
        }
      }
    }
  }
}
