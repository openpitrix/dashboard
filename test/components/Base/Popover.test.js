import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';

import Popover from 'components/Base/Popover';

describe('Base/Tooltip', () => {
  it('basic render', () => {
    const wrapper = render(
      <Popover className="test" content="test">
        hover me
      </Popover>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
