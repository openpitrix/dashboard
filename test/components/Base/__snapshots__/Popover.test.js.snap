// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Tooltip basic render 1`] = `
<div
  class="tooltip popover test"
>
  <div
    class="target"
  >
    hover me
  </div>
  <div
    class="popper pi-popover-popper"
    style="position:absolute;pointer-events:none;opacity:0"
  >
    <div>
      <div
        class="pi-popover-content"
      >
        test
      </div>
    </div>
  </div>
</div>
`;
